README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
Overview of the package
-----------------------

Basic example
-------------

You can use the `get_county_cd` function to get a vector of all counties within a state:

``` r
get_county_cd(state = c("Georgia", "Alabama"))
#>   [1] "01001" "01003" "01005" "01007" "01009" "01011" "01013" "01015"
#>   [9] "01017" "01019" "01021" "01023" "01025" "01027" "01029" "01031"
#>  [17] "01033" "01035" "01037" "01039" "01041" "01043" "01045" "01047"
#>  [25] "01049" "01051" "01053" "01055" "01057" "01059" "01061" "01063"
#>  [33] "01065" "01067" "01069" "01071" "01073" "01075" "01077" "01079"
#>  [41] "01081" "01083" "01085" "01087" "01089" "01091" "01093" "01095"
#>  [49] "01097" "01099" "01101" "01103" "01105" "01107" "01109" "01111"
#>  [57] "01113" "01115" "01117" "01119" "01121" "01123" "01125" "01127"
#>  [65] "01129" "01131" "01133" "13001" "13003" "13005" "13007" "13009"
#>  [73] "13011" "13013" "13015" "13017" "13019" "13021" "13023" "13025"
#>  [81] "13027" "13029" "13031" "13033" "13035" "13037" "13039" "13043"
#>  [89] "13045" "13047" "13049" "13051" "13053" "13055" "13057" "13059"
#>  [97] "13061" "13063" "13065" "13067" "13069" "13071" "13073" "13075"
#> [105] "13077" "13079" "13081" "13083" "13085" "13087" "13089" "13091"
#> [113] "13093" "13095" "13097" "13099" "13101" "13103" "13105" "13107"
#> [121] "13109" "13111" "13113" "13115" "13117" "13119" "13121" "13123"
#> [129] "13125" "13127" "13129" "13131" "13133" "13135" "13137" "13139"
#> [137] "13141" "13143" "13145" "13147" "13149" "13151" "13153" "13155"
#> [145] "13157" "13159" "13161" "13163" "13165" "13167" "13169" "13171"
#> [153] "13173" "13175" "13177" "13179" "13181" "13183" "13185" "13187"
#> [161] "13189" "13191" "13193" "13195" "13197" "13199" "13201" "13205"
#> [169] "13207" "13209" "13211" "13213" "13215" "13217" "13219" "13221"
#> [177] "13223" "13225" "13227" "13229" "13231" "13233" "13235" "13237"
#> [185] "13239" "13241" "13243" "13245" "13247" "13249" "13251" "13253"
#> [193] "13255" "13257" "13259" "13261" "13263" "13265" "13267" "13269"
#> [201] "13271" "13273" "13275" "13277" "13279" "13281" "13283" "13285"
#> [209] "13287" "13289" "13291" "13293" "13295" "13297" "13299" "13301"
#> [217] "13303" "13305" "13307" "13309" "13311" "13313" "13315" "13317"
#> [225] "13319" "13321"
```

You can use the `get_gages` function to pull all gages within a county or counties. For example, to get information on all gages for Miami-Dade county, you can run:

``` r
library(dplyr)
get_gages("12086", start_date = "1988-01-01", end_date = "2015-01-01") %>%
  slice(1:5)
#>   agency_cd         site_no
#> 1      USGS       022908295
#> 2      USGS 251203080480600
#> 3      USGS 251154080471900
#> 4      USGS 251033080440800
#> 5      USGS        02289040
#>                                          station_nm site_tp_cd dec_lat_va
#> 1 BOTTLE CREEK AT ROOKERY BRANCH NEAR HOMESTEAD, FL         ST   25.46798
#> 2   WEST LAKE OUTLET TO LONG LAKE NEAR FLAMINGO, FL         ST   25.20072
#> 3            CUTHBERT LAKE OUTLET NEAR FLAMINGO, FL         ST   25.19836
#> 4                    OYSTER CREEK NEAR FLAMINGO, FL         ST   25.17589
#> 5  TAMIAMI C OUTLETS L67A TO 40 MI BND NR MIAMI, FL         ST   25.76205
#>   dec_long_va county_cd DA
#> 1   -80.85453     12086 NA
#> 2   -80.80161     12086 NA
#> 3   -80.78867     12086 NA
#> 4   -80.73558     12086 NA
#> 5   -80.72590     12086 NA
```

You can use these two functions together within a pipe chain. For example, to get information on all the gages in Virginia in 2015, you can run:

``` r
va_gages <- get_county_cd("Virginia") %>%
   get_gages(start_date = "2015-01-01", end_date = "2015-12-31")
head(va_gages)
#>   agency_cd  site_no                                     station_nm
#> 1      USGS 03207800                    LEVISA FORK AT BIG ROCK, VA
#> 2      USGS 02013000                DUNLAP CREEK NEAR COVINGTON, VA
#> 3      USGS 02014000                 POTTS CREEK NEAR COVINGTON, VA
#> 4      USGS 02018500                 CATAWBA CREEK NEAR CATAWBA, VA
#> 5      USGS 02011800 JACKSON RIVER BL GATHRIGHT DAM NR HOT SPGS, VA
#> 6      USGS 02055100                TINKER CREEK NEAR DALEVILLE, VA
#>   site_tp_cd dec_lat_va dec_long_va county_cd    DA
#> 1         ST   37.35372   -82.19569     51027 297.0
#> 2         ST   37.80290   -80.04700     51005 162.0
#> 3         ST   37.72901   -80.04228     51005 153.0
#> 4         ST   37.46819   -80.00532     51023  34.3
#> 5         ST   37.94846   -79.94922     51005 345.0
#> 6         ST   37.41763   -79.93532     51023  11.7
```

The function sends county-by-county requests to the USGS water services server (<https://waterservices.usgs.gov>) and, if there are data for gages within that county, returns that data and adds it to the dataframe returned by the function. This function may result in a message about a 404 HTTP status for some counties. This means that no data was found for one of the counties being queried but will not affect proper running of the function for other counties. Internally, this function uses the `whatNWISsites` function from the package, `dataRetrieval`, but adds the county FIPS code for each gage, allowing users to join gage data with other data identified by county FIPS. This function also uses the `readNWISsites` function from the `dataRetrieval` package to obtain the drainage area of the watershed at the selected gage. This can be used as a proxy for relative river size.

Once you have a list of gage numbers for which you would like to pull flow data, you can use the `get_flow_data` function to pull that data. For example, to pull data for all the stream gage flow data for 1999 for gages in Southhampton County, VA (FIPS code 51175), you can run:

``` r
southampton_gages <- get_gages("51175", start_date = "1999-01-01",
                                end_date = "1999-12-31")
southampton_data <- get_flow_data(southampton_gages, start_date = "1999-01-01",
                                  end_date = "1999-12-31")
```

The output from `get_flow_data` is a data frame with daily streamflow for the queried gages over the requested dates. The data frame has three columns: "site\_no" = gage number, "date" = date of observation, and "discharge" = mean daily flow in cubic feet per second (USGS parameter code 00060):

``` r
head(southampton_data)
#>    site_no       date discharge
#> 1 02047000 1999-01-01       685
#> 2 02047000 1999-01-02       631
#> 3 02047000 1999-01-03       714
#> 4 02047000 1999-01-04      1300
#> 5 02047000 1999-01-05      2190
#> 6 02047000 1999-01-06      2830
```

``` r
library(ggplot2)
ggplot(southampton_data, aes(x = date, y = discharge, color = site_no)) + 
  geom_line() + theme_classic() + 
  labs(y = "mean daily flow (cubic ft / s)", color = "gage site #")
```

![](README-unnamed-chunk-8-1.png)

To determine when a flood occurred, we need a value to determine a gage-specific value of streamflow that consitutes a "flood". Expected streamflow values vary substantially from gage to gage, so it is criticial to determine a sensible threshold for each gage included in a study before trying to identify floods at each gage.

One way to get gage-specific flood thresholds is with the `find_nws` function, which gets the National Weather Service flood discharge threshold, when available, for a gage site. Note there are four NWS thresholds for each gage, representing cutoffs for different levels of floods: "Action", "Flood", "Moderate", and "Major". These thresholds are available through the Advanced Hydrologic Prediction Service's "River Stage" information. They are originally in feet for stages, which we convert to streamflow (in cubic feet per second) using USGS rating tables for each gage (pulled using the `readNWISrating` function from the `dataRetrieval` package).

Gages may have some, all, or none of these NWS flood thresholds available. If a gage does not have any of the four thresholds, it is excluded from the outpout of the `find_nws` function. For example, to get "Moderate" flood thresholds for the Virginia gages, you can run:

``` r
va_nws <- find_nws(site_no = va_gages$site_no, type = "moderate")
head(va_nws)
#>    site_no flood_val
#> 1 01652500  8400.246
#> 2 03168000 55378.500
#> 3 02060500 28306.221
#> 4 01654000  3370.084
#> 5 03173000 17958.120
#> 6 02011400  8008.627
summary(va_nws$flood_val)
#>    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#>    1719    8062   14040   22290   28310   90490
```

Many USGS gages do not have NWS flood thresholds for any of the four categories of floods. For example, out of the 171 gages pulled for Virginia for `va_gages`, only 101 have one or more NWS flood thresholds. Therefore, use of the NWS flood thresholds to identify floods may severely limit the sample size of the data output.

Another way to get gage-specific flood thresholds is with the `find_q2` function, which calculates the median annual flood for each gage using a minimum of 20 years of USGS annual peak flow data:

``` r
va_q2 <- find_q2(site_no = va_gages$site_no)
head(va_q2)
#> # A tibble: 6 × 3
#>    site_no flood_val years
#>      <chr>     <dbl> <int>
#> 1 01613900     923.0    56
#> 2 01615000    2355.0    68
#> 3 01620500     691.5    70
#> 4 01621050     543.0    23
#> 5 01622000    7440.0    87
#> 6 01625000    5740.0    89
summary(va_q2$flood_val)
#>    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#>   113.5  2606.0  4970.0  9441.0  9477.0 70770.0
```

For the Q2 flood threshold, 146 of the gages in Virginia have flood threshold values.

You can compare the results from these two methods for sites where you can get both values (note that both axes are shown on a log-10 scale):

``` r
va_flood_stage <- va_nws %>%
  rename(flood_nws = flood_val) %>%
  inner_join(va_q2, by = "site_no") %>%
  rename(flood_q2 = flood_val)
ggplot(va_flood_stage, aes(x = flood_q2, y = flood_nws)) + 
  geom_point(alpha = 0.5) + 
  geom_abline(aes(intercept = 0, slope = 1), linetype = 3) + 
  geom_smooth(method = "lm", se = FALSE) + 
  scale_x_log10(labels = scales::comma) + scale_y_log10(labels = scales::comma) + 
  labs(x = "Streamflow threshold for\nflood based on USGS Q2",
       y = "Streamflow threshold for flood based\non NWS flood height thresholds") + 
  theme_classic() + 
  expand_limits(x = c(min(va_flood_stage$flood_q2, va_flood_stage$flood_nws),
                      max(va_flood_stage$flood_q2, va_flood_stage$flood_nws)),
                y = c(min(va_flood_stage$flood_q2, va_flood_stage$flood_nws),
                      max(va_flood_stage$flood_q2, va_flood_stage$flood_nws)))
```

<img src="README-unnamed-chunk-11-1.png" style="display: block; margin: auto;" />

For the Virginia monitors, you can see that the flood values from the "moderate" NWS flood heights and the Q2 method are well-correlated, although NWS values tend to be consistently higher than Q2 values.

Whichever flood threshold you pick, it can be joined to the time series streamflow data obtained with the `get_flow_data` function, and flood status can be determined on each day based on whether streamflow that day exceeded the threshold. For example, the following code can be used to add a binary `flood` variable to the streamflow data pulled earlier for Southampton County, VA:

``` r
southampton_q2 <- find_q2(site_no = southampton_gages$site_no)
southampton_data <- southampton_data %>% 
  left_join(southampton_q2, by = "site_no") %>% 
  select(-years) %>% 
  mutate(flood = discharge >= flood_val)
head(southampton_data)
#>    site_no       date discharge flood_val flood
#> 1 02047000 1999-01-01       685      7980 FALSE
#> 2 02047000 1999-01-02       631      7980 FALSE
#> 3 02047000 1999-01-03       714      7980 FALSE
#> 4 02047000 1999-01-04      1300      7980 FALSE
#> 5 02047000 1999-01-05      2190      7980 FALSE
#> 6 02047000 1999-01-06      2830      7980 FALSE
```

Here are the time series for the two streamgages in the county, with horizontal lines added for the flood threshold used for each gage:

``` r
library(ggplot2)
ggplot(southampton_data, aes(x = date, y = discharge)) + 
  geom_line() + theme_classic() + 
  geom_point(aes(color = flood), alpha = 0.5, size = 1.2) + 
  facet_wrap(~ site_no, ncol = 1) +
  geom_hline(data = southampton_q2, aes(yintercept = flood_val),
             linetype = 3) +
  labs(y = "mean daily flow (cubic ft / s)", color = "Flood status")
```

![](README-unnamed-chunk-13-1.png)

Once you have data on gages, flood values, and flow data, you can also get flood summaries by site over the selected date range using the `flood_analysis` function. For each site, this function calculates the average "peak" (ratio of observed discharge to flood threshold discharge), maximum peak, and number of days where a flood occured. It also classifies flood magnitude. If the NWS flood threshold is used, this is a simple binary of "Flood" or "No Flood". If the Q2 flood threshold is used, flood magnitude is classified based on "max\_peak" value: "None" (&lt;1), "Minor" (1-1.5), "Moderate" (1.5-2), "Major" (2-5), and "Extreme" (&gt;5). For example, to get flood summary statistics by gage for the Virginia gages, you can run:

``` r
va_counties <- get_county_cd("Virginia")
va_flow_data <- get_flow_data(va_gages, start_date = "2015-01-01",
                              end_date = "2015-12-31")
va_floods <- flood_analysis(flow_data = va_flow_data, peaks = va_q2, 
                            gages = va_gages, county_cd = va_counties,
                            threshold = "Q2")
head(va_floods, 3)
#> # A tibble: 3 × 14
#>    site_no county_cd      lat      long   avg_peak flood_dur max_peak
#>      <chr>     <chr>    <dbl>     <dbl>      <dbl>     <int>    <dbl>
#> 1 02072000     51067 36.78069 -80.02476 0.26734310         8 3.953871
#> 2 02040000     51007 37.42154 -77.85889 0.12193978         5 1.880193
#> 3 03524500     51195 36.92927 -82.45626 0.06564404         2 1.876866
#> # ... with 7 more variables: num_missing <int>, Q2 <lgl>, DA <dbl>,
#> #   size <dbl>, state <chr>, county <chr>, flood <fctr>
```

``` r
ggplot(va_floods, aes(x = max_peak, fill = flood)) + 
  geom_histogram()
#> Warning: Removed 31 rows containing non-finite values (stat_bin).
```

<img src="README-unnamed-chunk-15-1.png" style="display: block; margin: auto;" />

There is also a function that will allow you to get county-level aggregate statistics from this gage-level summary. Any counties with no gages or for which the gages didn't have flow data or flood threshold values are also included. If the Q2 flood threshold is used, this function gives the percentage of gages in the county in above each flood magnitude classification (e.g. "Minor", "Moderate", etc.):

``` r
va_county_stats <- county_aggregates(flood_stats = va_floods)
head(va_county_stats)
#> # A tibble: 6 × 12
#>   county_cd        county    state num_gage  avg_peak  max_peak minor
#>       <chr>         <chr>    <chr>    <int>     <dbl>     <dbl> <dbl>
#> 1     51001 accomack:main virginia        2       NaN        NA    -1
#> 2     51003     albemarle virginia        4 0.3655210 0.5557300     0
#> 3     51005     alleghany virginia        4 0.7488700 0.8574018     0
#> 4     51007        amelia virginia        2 1.5193316 1.8801932   100
#> 5     51009       amherst virginia        1       NaN        NA    -1
#> 6     51011    appomattox virginia        1 0.3571429 0.3571429     0
#> # ... with 5 more variables: moderate <dbl>, major <dbl>, extreme <dbl>,
#> #   max_dur <int>, avg_dur <dbl>
```


NEWS.md
# countyfloods 0.1.0
This update fixes some bugs and allows data retrieval and analysis for Puerto Rico,
Hawaii, and Alaska. Also, I added a function to map a series of days of flood results.

# countyfloods 0.0.2
This update retains compatability with newer versions of the dataRetrieval package.

# countyfloods 0.0.1
This is the initial release of this package.


tests/testthat.R
library(testthat)
library(countyfloods)

test_check("countyfloods")


R/map_results.R
#' Function maps data, either by gage or by county
#'
#' Displays a state or multi-state map summarizing flood analysis results either
#' by gage or county.
#'
#' @param flood_stats Either a data frame of flood analysis results, by gage or by county,
#'   or a list of both data frames.
#' @param category Character string of the flood magnitude category to be used
#'   for mapping (one of "minor", "moderate", "major", or "extreme"). This
#'   parameter only works when mapping county-level, rather than gage-level,
#'   values.
#'
#' @return A map of the state(s) analyzed showing counties and gages color coded
#'   based on flood magnitude, depending on the type of data in flood_stats. Gage
#'   flood thresholds are "None" (flood_ratio < 1), "Minor" (flood_ratio < 1.5),
#'   "Moderate" (flood_ratio < 2), "Major" (flood_ratio < 5), and "Extreme"
#'   (flood_ratio > 5). For county aggregate maps, flood exposure is assessed based
#'   on the percentage of gages in the county at or above a specified flood threshold.
#'   Exposure categories include "Low" (0% - 20%), "Moderate" (20% - 40%), "Moderate-High"
#'   (40% - 60%), "High" (60% - 80%), and "very High" (80% - 100%).
#'
#' @examples
#' \dontrun{
#' #Use Q2 as flood threshold and get get gage-level output
#' va_floods <- run_flood(state = "Virginia", start_date = "2015-01-01",
#'                       end_date = "2015-12-31", threshold = "Q2", output = "gage")
#' #Map results by gage
#' map_flood(va_floods)
#'
#'
#' #Use NWS flood thresholds and get county-level output
#' va_floods <- run_flood(state = "Virginia", start_date = "2015-01-01",
#'                       end_date = "2015-12-31", threshold = "NWS", flood_type = "action",
#'                       output = "county")
#' #Map results by county
#' map_flood(va_floods)
#' }
#' @export
map_flood <- function(flood_stats, category = "minor") {

  #Check if flood_stats data is at the gage- or county-level and call
  #appropriate mapping function
  if (!is.data.frame(flood_stats)) {
    output <- "both"
  }else if (names(flood_stats)[1] == "site_no") {
    output <- "gage"
  }else if (names(flood_stats)[1] == "county_cd") {
    output <- "county"
  }

  if (output == "gage") {
    map_gage(flood_stats)
  }else if (output == "county") {
    map_county(flood_stats, category = category)
  }else if (output == "both") {
    gage_map <- map_gage(flood_stats[[1]])
    county_map <- map_county(flood_stats[[2]], category = category)
    return(list(gage_map, county_map))
  }
}

#' Maps flood data by gage
#'
#' Creates a state level map of flood analysis output by USGS gage. Gages are
#' color coded based on maximum flood magnitude (for flood threshold = "Q2").
#' If the flood threshold is "NWS", points are binary coded based on flood
#' occurrence (e.g. yes/no).
#'
#' @param flood_stats Data frame of flood analysis results, by gage.
#' @param date Date of data mapped to be printed at the top of the image. This
#'   argument is used by the `time_series_map` function.
#'
#' @return A map of USGS gages color coded by maximum flood magnitude.
#'
#' @importFrom dplyr %>%
#'
#' @export
map_gage <- function(flood_stats, date = "") {

  if (dplyr::first(flood_stats$flood) == "No Flood" | dplyr::first(flood_stats$flood) == "Flood"){
    colors <- colors <- c("#993404", "#FFFFFF")
    names(colors) <- c("Flood", "No Flood")
  } else {
    colors <- c("#993404", "#D95F0E", "#FE9929", "#FED98E", "#FFFFFF")
    names(colors) <- c("Extreme", "Major", "Moderate", "Minor", "None")
  }

  #Check "size" variable, if NA, use other metric (Q2 or DA)
  if (sum(!is.na(flood_stats$size)) == 0 && sum(!is.na(flood_stats$Q2)) == 0) {
    #Q2 NA, replace size with DA
    flood_stats <- flood_stats %>%
      dplyr::mutate_(size = ~ replace(size, is.na(size), log10(DA)))
    warning("Point size based on log10(DA).")
  }else if(sum(!is.na(flood_stats$size)) == 0 && sum(!is.na(flood_stats$Q2)) == 0) {
    #DA NA, replace size with Q2
    flood_stats <- flood_stats %>%
      dplyr::mutate_(size = ~ replace(size, is.na(size), log10(Q2)))
    warning("Point size based on log10(Q2).")
  }

  #sort to print highest floods last
  if ("flood_ratio" %in% colnames(flood_stats)){
    flood_stats <- flood_stats %>%
      dplyr::arrange_(~ flood_ratio)
  }else{
    flood_stats <- flood_stats %>%
      dplyr::arrange_(~ max_peak)
  }

  if (!(dplyr::first(flood_stats$state) %in% c("puerto rico", "hawaii", "alaska"))){
    region <- as.character(unique(flood_stats$state))

    counties <- ggplot2::map_data("county", region = region)
    counties_sub <- counties[counties$subregion %in% flood_stats$county[!is.na(flood_stats$lat)], ]
    counties_sub_ND <- counties[counties$subregion %in% flood_stats$county[is.na(flood_stats$lat)], ]
    ggplot2::ggplot(counties_sub, ggplot2::aes_string(x = "long", y = "lat", group = "group")) +
      ggplot2::geom_polygon(fill = "gray95", color = "black") +
      ggplot2::geom_polygon(data = counties, ggplot2::aes_string(x = "long", y = "lat", group = "group"),
                            fill = NA, color = "black") +
      ggplot2::geom_polygon(data = counties_sub_ND, ggplot2::aes_string(x = "long", y = "lat", group = "group"),
                            fill = "gray60", color = "black") +
      ggplot2::geom_point(data = flood_stats, ggplot2::aes_string(x = "long", y = "lat", group = NA, fill = "flood",
                          size = "size"), alpha = 0.8, pch = 21, show.legend = TRUE) +
      ggplot2::guides(size = "none") +
      ggplot2::scale_fill_manual(values = colors, drop = FALSE) +
      ggplot2::coord_map() +
      ggplot2::theme_void() +
      ggplot2::labs(title = date) +
      ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5))
  }
  else{
    if (dplyr::first(flood_stats$state) == "puerto rico"){
      state <- ggplot2::map_data("world", region = "Puerto Rico")
    }
    else if (dplyr::first(flood_stats$state) == "alaska"){
      state <- ggplot2::map_data("world", xlim = c(-170, -130), ylim = c(50, 72), lforce = "e")
    }
    else {
      state <- ggplot2::map_data("world", xlim = c(-162, -154), ylim = c(18, 23))
    }


    ggplot2::ggplot(state, ggplot2::aes_string(x = "long", y = "lat", group = "group")) +
      ggplot2::geom_polygon(data = state, ggplot2::aes_string(x = "long", y = "lat", group = "group"),
                            fill = NA, color = "black") +
      ggplot2::geom_point(data = flood_stats, ggplot2::aes_string(x = "long", y = "lat", group = NA, fill = "flood",
                                                                  size = "size"), alpha = 0.8, pch = 21, show.legend = TRUE) +
      ggplot2::guides(size = "none") +
      ggplot2::scale_fill_manual(values = colors, drop = FALSE) +
      ggplot2::coord_map() +
      ggplot2::theme_void() +
      ggplot2::labs(title = date) +
      ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5))
  }
}

#' Maps flood data by county
#'
#' Creates a state level map of flood analysis output by county. Counties are
#' color coded based on the percentage of gages in that county at or exceeding a
#' given flood magnitude.
#'
#' @param county_stats Data frame of flood analyasis results, summarized by
#'   county.
#' @param category Character string of the flood magnitude category to be used
#'   for mapping (one of "minor", "moderate", "major", or "extreme"). This
#'   parameter only works when mapping county-level, rather than gage-level,
#'   values.
#' @param date Date of data mapped to be printed at the top of the image. This
#'   argument is used by the `time_series_map` function.
#'
#' @return A map of counties color coded by percentage of gages experiencing flooding.
#'
#' @importFrom dplyr %>%
#'
#' @export
map_county <- function(county_stats, category = "minor", date = "") {

  if (county_stats$state %in% c("puerto rico", "alaska", "hawaii")){
    stop("Sorry but we can't produce county maps for Puerto Rico, Alaska, or Hawaii.")
  }

  #Check inputs and return error messages as necessary
  category <- tolower(category)
  if(category != "minor" & category != "moderate" & category != "major" &
     category != "extreme") stop("Input category must be one of 'minor', moderate', 'major', or 'extreme'")

  colors <- c("#993404", "#D95F0E", "#FE9929", "#FED98E", "#FFFFD4", "#bfbfbf")
  exposure_cat <- c("Very High", "High", "Moderate-High", "Moderate", "Low", "No Data")
  names(colors) <- exposure_cat

  #Set flood exposure "categories" based on percentage of flooded gages and the
  #user-defined metric (e.g. minor, major, extreme flooding)
  if ("no_flood" %in% colnames(county_stats)){
    county_stats <- county_stats %>%
      dplyr::select_(.dots = list("county", "state", "no_flood",
                                  "yes_flood")) %>%
      tidyr::gather_(key_col = "key", value_col = "value",
                     gather_cols = c("no_flood", "yes_flood")) %>%
      dplyr::filter_(~ key == "yes_flood") %>%
      dplyr::mutate_(cat = ~ cut(value, breaks = c(-1, 0, 20, 40, 60, 80, 100),
                                 labels = c("No Data", "Low", "Moderate", "Moderate-High",
                                            "High", "Very High"),
                                 include.lowest = TRUE, right = FALSE))
  }else {
    county_stats <- county_stats %>%
      dplyr::select_(.dots = list("county", "state", "minor",
                                  "moderate", "major", "extreme")) %>%
      tidyr::gather_(key_col = "key", value_col = "value",
                    gather_cols = c("minor", "moderate", "major", "extreme")) %>%
      dplyr::filter_(~ key == category) %>%
      dplyr::mutate_(cat = ~ cut(value, breaks = c(-1, 0, 20, 40, 60, 80, 100),
                              labels = c("No Data", "Low", "Moderate", "Moderate-High",
                                         "High", "Very High"),
                              include.lowest = TRUE, right = FALSE))
  }

  region <- as.character(unique(county_stats$state))

  #Get all counties for states analyzed as well as the subset of counties with
  # actual data
  counties <- ggplot2::map_data("county", region = region)
  counties_sub <- counties[counties$subregion %in% county_stats$county, ]

  counties_sub$cat <- county_stats$cat[match(counties_sub$subregion,
                                             county_stats$county)]

  ggplot2::ggplot(counties_sub, ggplot2::aes_string(x = "long", y = "lat", group = "group")) +
    ggplot2::geom_polygon(ggplot2::aes_string(fill = "cat"), color = "black") +
    ggplot2::geom_polygon(data = counties, ggplot2::aes_string(x = "long", y = "lat", group = "group"),
                          fill = NA, color = "black") +
    ggplot2::scale_fill_manual(values = colors, drop = FALSE) +
    ggplot2::coord_map() +
    ggplot2::theme_void() +
    ggplot2::labs(title = date) +
    ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5))
}

#' Function plots time series data by county
#'
#' Displays four time series bar charts per county displaying the number of
#' gages with flooding, maximum flood ratio, average flood ratio, and the
#' percent of gages above a specified flood threshold.
#'
#' @param county_series Data frame of flood time series results by county,
#'   output of \code{time_series_flood} function.
#' @param category Character string of the flood magnitude category to be used
#'   for mapping (one of "minor", "moderate", "major", or "extreme").
#' @param start_date Character string of start date for x-axis of plots. If not
#'   specified, defaults to the earliest observed flood in the data.
#' @param end_date Character string of end date for x-axis of plots. If not
#'   specified, defaults to the latest observed flood in the data.
#'
#' @return Four time series bar charts per county displaying the number of
#' gages with flooding, maximum flood ratio, average flood ratio, and the
#' percent of gages above a specified flood threshold.
#'
#' @examples
#' \dontrun{
#' #Use Q2 as flood threshold
#' va_time_series <- time_series_flood(state = "Virginia", start_date = "2015-01-01",
#'                       end_date = "2015-12-31", threshold = "Q2")
#'
#' #Map results
#' time_series_plot(va_time_series[[2]])
#' }
#'
#' @importFrom grDevices dev.off png
#'
#' @export
time_series_plot <- function(county_series, category = "moderate",
                             start_date = NULL, end_date = NULL) {

  if(is.null(start_date)) {start_date <- min(county_series$date)}
  if(is.null(end_date)) {end_date <- max(county_series$date)}

  if ("no_flood" %in% colnames(county_series)){
    category = "yes_flood"
  }

  no_output <- suppressWarnings(plyr::ddply(county_series, "county", function(x) {

  p1 <- ggplot2::ggplot(data = x, ggplot2::aes_string(x = "date", y = "num_gage")) +
    ggplot2::geom_bar(stat = "identity", width = 10) +
    ggplot2::xlim(start_date, end_date) +
    ggplot2::ylab("Gages") +
    ggplot2::xlab("Date") +
    ggplot2::ggtitle(substitute(paste("Number of gages with a flood (", county, " County)"),
                                list(county = R.utils::capitalize(unique(x$county))))) +
    ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5))

  p2 <- ggplot2::ggplot(data = x, ggplot2::aes_string(x = "date", y = "max_peak")) +
    ggplot2::geom_bar(stat = "identity", width = 10) +
    ggplot2::xlim(start_date, end_date) +
    ggplot2::ylab("Flood Ratio") +
    ggplot2::xlab("Date") +
    ggplot2::ggtitle("Maximum flood ratio") +
    ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5))

  p3 <- ggplot2::ggplot(data = x, ggplot2::aes_string(x = "date", y = "avg_peak")) +
    ggplot2::geom_bar(stat = "identity", width = 10) +
    ggplot2::xlim(start_date, end_date) +
    ggplot2::ylab("Flood Ratio") +
    ggplot2::xlab("Date") +
    ggplot2::ggtitle("Average flood ratio") +
    ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5))

  p4 <- ggplot2::ggplot(data = x, ggplot2::aes_(x = ~ date, y = ~ x[ ,tolower(category)])) +
    ggplot2::geom_bar(stat = "identity", width = 10) +
    ggplot2::xlim(start_date, end_date) +
    ggplot2::ylab("% Above") +
    ggplot2::xlab("Date") +
    ggplot2::ggtitle("Percent of gages above flood threshold") +
    ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5))

  grid::grid.newpage()
  grid::grid.draw(rbind(ggplot2::ggplotGrob(p1), ggplot2::ggplotGrob(p2), ggplot2::ggplotGrob(p3),
                        ggplot2::ggplotGrob(p4), size = "last"))

  return(NA)
  }))

}

#' Function maps time series flood data, either by gage or by county
#'
#' Displays a state or multi-state map summarizing flood analysis results either
#' by gage or county for each individual date with data.
#'
#' @param flood_stats Either a data frame of flood analysis results, by gage or by county,
#'   or a list of both data frames.
#' @param category Character string of the flood magnitude category to be used
#'   for mapping (one of "minor", "moderate", "major", or "extreme"). This
#'   parameter only works when mapping county-level, rather than gage-level,
#'   values.
#' @param filename Character string of the file path and beginning of name of where
#'   to save the individual images generated. For example, "C:/Desktop/TX" would produce
#'   image files saved as "TX_Gage_Map_Date.png" or "TX_County_Map_Date.png" where
#'   "Date" is replaced by the actual date. If no filename is provided, the images
#'   aren't saved but are displayed in the plot viewer.
#'
#' @return For each date with data, a map of the state(s) analyzed showing counties
#'   and gages color coded based on flood magnitude, depending on the type of data
#'   in flood_stats. Gage flood thresholds are "None" (flood_ratio < 1), "Minor"
#'   (flood_ratio < 1.5), "Moderate" (flood_ratio < 2), "Major" (flood_ratio < 5),
#'   and "Extreme" (flood_ratio > 5). For county aggregate maps, flood exposure
#'   is assessed based on the percentage of gages in the county at or above a specified
#'   flood threshold. Exposure categories include "Low" (0% - 20%), "Moderate"
#'   (20% - 40%), "Moderate-High" (40% - 60%), "High" (60% - 80%), and "very High"
#'   (80% - 100%).
#'
#' @examples
#' \dontrun{
#' #Use the time_series_flood function with option filter_data = FALSE to get data
#'   tx <- time_series_flood(state = "Texas", start_date = "2017-08-24", end_date =
#'         "2017-09-10", filter_data = FALSE)
#'
#' #Returns a list of gage data and county data. Can map them individually or all at once
#    time_series_map(tx)
#' }
#' @export
time_series_map <- function(flood_stats, category = "minor", filename = ""){

  #Check if flood_stats data is at the gage- or county-level and call
  #appropriate mapping function
  if (!is.data.frame(flood_stats)) {
    output <- "both"
  }else if (names(flood_stats)[1] == "site_no") {
    output <- "gage"
  }else if (names(flood_stats)[1] == "county_cd") {
    output <- "county"
  }

  if (output == "gage") {
    plyr::d_ply(flood_stats, "date", function(x, filename){
      if (filename != ""){
        png(paste0(filename, "_Gage_Map_", x$date[1], ".png"), type = "cairo", units = "in",
            height = 4, width = 4, res = 500)
      }
      print(map_gage(flood_stats = x, date = x$date[1]))

      if (filename != ""){
        dev.off()
      }

    }, filename)

  }else if (output == "county") {
    plyr::d_ply(flood_stats, "date", function(x, category, filename){
      if (filename != ""){
        png(paste0(filename, "_County_Map_", x$date[1], ".png"), type = "cairo", units = "in",
            height = 4, width = 4, res = 500)
      }
      print(map_county(county_stats = x, category = category, date = x$date[1]))

      if (filename != ""){
        dev.off()
      }
    }, category, filename)
  }else if (output == "both") {
    plyr::d_ply(flood_stats[[1]], "date", function(x){
      print(map_gage(flood_stats = x, date = x$date[1]))
    })
    plyr::d_ply(flood_stats[[2]], "date", function(x){
      print(map_county(county_stats = x, category = category, date = x$date[1]))
    })
  }

}
